File: fastapi/api/controller/uploadController.py
from fastapi import APIRouter, Depends, File, Request, UploadFile
from starlette.concurrency import run_in_threadpool
from typing import Any, Dict, Optional
from api.service import AnalyzeService, get_analyze_service, UploadService, get_upload_service
from entity.dto import UploadDTO
from common.decorators import log, requireInternalToken
from common.utils import success

router: APIRouter = APIRouter(
    prefix="/upload",
    tags=["上传文件接口"],
)

@router.post(
    "",
    summary="上传文件",
    description="上传文件到OSS"
)
@requireInternalToken
@log("上传文件")
async def get_wordcloud(
    _: Request,
    data: UploadDTO, 
    analyzeService: AnalyzeService = Depends(get_analyze_service)
) -> Any:
    """上传文件接口"""
    
    oss_url: str = await run_in_threadpool(
        analyzeService.upload_file, 
        data.local_file, 
        data.oss_file
    )
    return success(oss_url)

@router.post(
    "/image",
    summary="上传图片",
    description="上传图片到OSS"
)
@log("上传图片")
async def upload_image_to_oss(
    _: Request, 
    file: UploadFile = File(...), 
    uploadService: UploadService = Depends(get_upload_service)
) -> Any:
    """上传图片接口"""
    
    file_dict: Dict[str, Any] = {
        "filename": file.filename,
        "file": file.file
    }
    result: Dict[str, Any] = await uploadService.handle_image_upload(file_dict)
    return success(result)

@router.post(
    "/pdf",
    summary="上传PDF",
    description="上传PDF到OSS"
)
@log("上传PDF")
async def upload_pdf_to_oss(
    _: Request, 
    file: UploadFile = File(...), 
    custom_filename: Optional[str] = None, 
    uploadService: UploadService = Depends(get_upload_service)
) -> Any:
    """上传PDF接口"""
    
    file_dict: Dict[str, Any] = {
        "filename": file.filename,
        "file": file.file
    }
    result: Dict[str, Any] = await uploadService.handle_pdf_upload(file_dict, custom_filename)
    return success(result)

File: nestjs/src/api/upload/upload.controller.ts
import { Body, Controller, Post, Query, Req } from '@nestjs/common';
import {
  ApiBody,
  ApiConsumes,
  ApiOperation,
  ApiQuery,
  ApiTags,
} from '@nestjs/swagger';
import type { FastifyRequest } from 'fastify';
import { BusinessException } from 'src/common/exceptions/business.exception';
import { Constants } from 'src/common/utils/constants';
import { ApiResponse, success } from 'src/common/utils/response';
import { ApiLog } from 'src/framework/decorators/apiLog.decorator';
import { RequireInternalToken } from 'src/framework/decorators/requireInternalToken.decorator';
import { UploadDto, UploadResult } from './dto/upload.dto';
import { MultipartUploadFile, UploadService } from './upload.service';

@Controller('upload')
@ApiTags('文件上传')
export class UploadController {
  constructor(private readonly uploadService: UploadService) {}

  @Post('')
  @ApiOperation({
    summary: '上传本地文件到 OSS',
    description: '通过本地文件路径上传文件到 OSS',
  })
  @RequireInternalToken()
  @ApiLog('上传文件到 OSS')
  async uploadFile(@Body() dto: UploadDto): Promise<ApiResponse<string>> {
    const url: string = await this.uploadService.uploadFile(
      dto.localFile,
      dto.ossFile,
    );
    return success(url);
  }

  @Post('image')
  @ApiConsumes('multipart/form-data')
  @ApiBody({
    schema: {
      type: 'object',
      properties: {
        file: {
          type: 'string',
          format: 'binary',
        },
      },
      required: ['file'],
    },
  })
  @ApiOperation({
    summary: '上传图片到 OSS',
    description: '通过 multipart/form-data 上传图片到 OSS',
  })
  @ApiLog('上传图片到 OSS')
  async uploadImage(
    @Req() req: FastifyRequest,
  ): Promise<ApiResponse<UploadResult>> {
    const parts = req.parts();
    for await (const part of parts) {
      if ('filename' in part && part.filename) {
        const result: UploadResult = await this.uploadService.uploadImage(
          part as MultipartUploadFile,
        );
        return success(result);
      }
    }
    throw BusinessException.notFound(Constants.FILE_PART_NOT_FOUND);
  }

  @Post('pdf')
  @ApiConsumes('multipart/form-data')
  @ApiBody({
    schema: {
      type: 'object',
      properties: {
        file: {
          type: 'string',
          format: 'binary',
        },
      },
      required: ['file'],
    },
  })
  @ApiQuery({
    name: 'custom_filename',
    type: 'string',
    required: false,
    description: '自定义文件名（不包含扩展名）',
  })
  @ApiOperation({
    summary: '上传 PDF 到 OSS',
    description: '通过 multipart/form-data 上传 PDF 到 OSS',
  })
  @ApiLog('上传 PDF 到 OSS')
  async uploadPdf(
    @Req() req: FastifyRequest,
    @Query('customFilename') customFilename?: string,
  ): Promise<ApiResponse<UploadResult>> {
    const parts = req.parts();
    for await (const part of parts) {
      if ('filename' in part && part.filename) {
        const result: UploadResult = await this.uploadService.uploadPdf(
          part as MultipartUploadFile,
          customFilename,
        );
        return success(result);
      }
    }
    throw BusinessException.notFound(Constants.FILE_PART_NOT_FOUND);
  }
}
